### ISSUE TITLE
NPM5 Should never assume and it's inconsistent

### ISSUE THREAD

**Issue by c0debreaker** [npm/npm#18617]
On a new project, I ran `npm install my-library@1.0.0`. The my-library is already in my locally installed Verdaccio(Sinopia). There are 2 versions I published, `1.0.0` and `1.0.1`.

After running the command above, this is my exercise app's package.json
```
{
  "name": "mainapp",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "my-library": "^1.0.0"
  }
}
```
This also created package-lock.json file. It contains the following
```
  "dependencies": {
    "my-library": {
      "version": "1.0.0",
      "resolved": "http://localhost:4873/my-library/-/my-library-1.0.0.tgz",
      "integrity": "sha512-J8Zp5/egXu/Qd2a1QE5AkCOzL8yQOHgHNqSwzSt13zn8PVN62rlioG+IPvqrFbnjVEViwM33PLImaY0JSEkgyA=="
    },
```
The my-library also got installed in node_modules and node_modules/my-library/package.json is showing version 1.0.0.

I ran `rm -rf node_modules` and I also ran `npm install`. The version of my-library that was installed in node_modules is correct. I ran it many times and version is correct.

However, if I execute `npm install my-library`, it installs the latest version from the registry and not from the entry in my package-lock.json. I was thinking that npm5 will use the version that it finds in package-lock.json but unfortunately not. It would be acceptable if the command was `npm install my-library@1.0.1`.

I really think NPM@5 should not assume.

I tried it on 5.3.0 and 5.4.2

**Comment by c0debreaker** [comment]
I also think it's inconsistent since NPM also has the `update` parameter which we can execute as `npm update` and it updates all the packages. If it was `npm update my-library` (which works) and not `npm install my-library`, it would have been clearer. From my testing, it looks like `npm install my-library` bypasses package-lock.json. So I can say that NPM is being inconsistent.

**Comment by c0debreaker** [comment]
I think this makes more sense

`npm install mylib` should install the exact version mentioned in package-lock.json. It should not bypass package-lock.json. If it's not in package-lock.json, consult package.json and use the semver defined and install it.

`npm install mylib@x.y.z` should bypass package-lock.json and package.json and should install x.y.z

`npm update mylib` should bypass package-lock.json and follow the value in package.json and use the latest

**Comment by Oliboy50** [comment]
agreed, exactly what does the PHP package manager => `composer`

but `npm shrinkwrap` is almost exactly what you're looking for @c0debreaker :)

**Comment by c0debreaker** [comment]
Yup, I quite agree with you on npm shrinkwrap. I've been asking myself over the weekend "What was package-lock.json made for, what is the use of it?"

**Comment by ljharb** [comment]
package-lock can't be published; npm-shrinkwrap can.

Publishing a lockfile prevents consumers from deduping your module's dep graph, so it's pretty user-hostile. In a top-level app (not published) then either choice is fine, except it seems slightly more ergonomic to work with package-lock than shrinkwrap.

**Comment by mk-pmb** [comment]
> npm install mylib should install the exact version mentioned in package-lock.json.

Why would you need to issue that command in the first place? If mylib is a dependency, you wouldn't have to explicitly install it. If it's not a dep, why is it in package-lock.json?

**Comment by c0debreaker** [comment]
It's in dependency. It's our internal library which contains components. It is in package-lock.json because npm5 added it.

I see what you're saying. If we wanted to update mylib alone and not others, how would you do it? We should execute `npm install mylib` right? However, few days ago, my impression on `npm install mylib` was that npm will install the version mentioned in package-lock.json. However, after trying it a lot, `npm install mylib` will definitely grab the latest from registry and ignore package-lock.json which is fine. It's just getting used to it.

I also don't want to run `npm update` which updates everything. I'd rather execute
`npm install mylib1@latest mylib2@latest mylib3@latest mylib4@latest mylib5@latest mylib6@latest`

**Comment by mk-pmb** [comment]
> If we wanted to update mylib alone and not others, how would you do it?

In simple scenarios, you can update the version range in the dependency declaration in `package.json`, then run `npm i .` (in your module's root directory) to have npm arrange a proper environment for your module. My expectation is that npm will determine and apply the minimum changes required to satisfy (1) your module's deps and (2) if possible, globally configured optimization goals for e.g. directory structure layout. (Edit: Implied: Thus, npm's actions might turn out to be just updating the package whose version range you changed.)

There may also be ways to restrict the scope of `npm update`. I guess it might be as easy as appending a package spec to the command.

Edit 4:

> I'd rather execute
> `npm install mylib1@latest mylib2@latest mylib3@latest mylib4@latest mylib5@latest mylib6@latest`

I'd expect that to work as well. Don't forget `--save` if you want your `package.json` updated.

**Comment by hagb4rd** [comment]
everything worked so well before the shrimshrim package-lock era. npm may have not solved all problems but was a decent tool i used to claim to be the best pm around.. its minimalist approach proved once more that simplicity is the result of maturity. so.. i want to go back.. anybody knows how far i need to go?
